### PR TITLE
✨ (go/v4): Add tests for the webhooks and controllers which are created so that suite_test are no longer ignored

### LINKED ISSUE #3646: BeforeSuite and AfterSuite modules not running in the ginkgo v2
### What broke? What's expected?

What broke? 
The BeforeSuite and AfterSuite modules not running in the ginkgo v2. In other words, the current `make test` do nothing in all project with the ginkgo v2.
Refer to the [ginkgo v2.0 note](https://onsi.github.io/ginkgo/MIGRATING_TO_V2#minor-additions-and-improvements), the BeforeSuite and AfterSuite will not run if all tests in a suite are skipped.

What's expected?
In the project with the ginkgo v2, the `make test` command can run successfully as expected and do the operations such as installing CRD actually.

### Reproducing this issue

1. Move into a test-project with ginkgo v1.
```bash
$ cd project
```

Add a wrong assertion as follows.
```yaml
var _ = BeforeSuite(func(done Done) {
    Expect(1).To(Equal(2))
}, 60)

var _ = AfterSuite(func() {
    Expect(1).To(Equal(2))
})
```

Run `make test`, and you will get some errors because the BeforeSuite module **executed actually**.

2. Move into a test-project with ginkgo v1.
```bash
$ cd project/testdata/project-v4
```

Add a wrong assertion as follows.
```yaml
var _ = BeforeSuite(func() {
    Expect(1).To(Equal(2))
})
```

Run `make test`, and you will run successfully because the BeforeSuite module **not executed actually**.

### KubeBuilder (CLI) Version

3.12.0

### PROJECT version

3

### Plugin versions

```yaml
# Code generated by tool. DO NOT EDIT.
# This file is used to track the info used to scaffold your project
# and allow the plugins properly work.
# More info: https://book.kubebuilder.io/reference/project-config.html
domain: testproject.org
layout:
- go.kubebuilder.io/v4
projectName: project-v4
repo: sigs.k8s.io/kubebuilder/testdata/project-v4
resources:
- api:
    crdVersion: v1
    namespaced: true
  controller: true
  domain: testproject.org
  group: crew
  kind: Captain
  path: sigs.k8s.io/kubebuilder/testdata/project-v4/api/v1
  version: v1
  webhooks:
    defaulting: true
    validation: true
    webhookVersion: v1
- api:
    crdVersion: v1
    namespaced: true
  controller: true
  domain: testproject.org
  group: crew
  kind: FirstMate
  path: sigs.k8s.io/kubebuilder/testdata/project-v4/api/v1
  version: v1
  webhooks:
    conversion: true
    webhookVersion: v1
- api:
    crdVersion: v1
  controller: true
  domain: testproject.org
  group: crew
  kind: Admiral
  path: sigs.k8s.io/kubebuilder/testdata/project-v4/api/v1
  plural: admirales
  version: v1
  webhooks:
    defaulting: true
    webhookVersion: v1
- controller: true
  domain: testproject.org
  group: crew
  kind: Laker
  version: v1
version: "3"
```

### Other versions

_No response_

### Extra Labels

_No response_

### PR DESCRIPTION
**Pull request by dashanji**
## How to reproduce?

Add any valid ginkgo assert such as `Expect(1).To(Equal(2))` to the BeforeSuite and  AfterSuite. As a result, the test should fail, but it succeeds actually.

## What do these changes do?

Add the unit test template for webhook and controllers so that suite_test are no longer ignored.

Fixes #3646

### CODE REVIEW

**Review by camilamacedo86 — CHANGES_REQUESTED** [review]
Tests should not be added to the suite_test by default.
The scaffold scaffolds the suite_test and not the tests for each API/controller and/or webhook

So, what is the use case that you would add assert checks to the Before/AfterSuite without properly implementing the tests for each controller and/or webhook to face the issue scenario described which you are trying to solve in this one?

**Comment by dashanji** [comment]
> Given this, I'm curious: under what circumstances would you add assert checks to the Before/AfterSuite without first implementing comprehensive tests for each controller and/or webhook, especially in light of the issue scenario you've described? Could you please provide more details on this?


Hi, @camilamacedo86, thanks for the details. Actually, in ginkgo v1, the BeforeSuite and AfterSuite can run successfully, so users can test the default case(CRDs, Controllers, and Webhooks can be installed successfully, just as what BeforeSuite does) without any changes just running `make test`. My point is to reserve the same adaption for users in the newest kubebuilder(v4). I don't know whether it makes sense to the community.

**Comment by camilamacedo86** [comment]
> Hi, @camilamacedo86, thanks for the details. Actually, in ginkgo v1, the BeforeSuite and AfterSuite can run successfully, so users can test the default case(CRDs, Controllers, and Webhooks can be installed successfully, just as what BeforeSuite does) without any changes just running make test. My point is to reserve the same adaption for users in the newest kubebuilder(v4). I don't know whether it makes sense to the community.

Can you describe what problem and when are you facing?
If I create a new project (kubebuilder init)
Scaffold an API (kubebuilder create api)
Run (make test)

All works fine. So, can you pelase raise an issue with:
- What are the steps that you are performing
- What would like to see
- What are you checking instead?

**Comment by dashanji** [comment]
> > Hi, @camilamacedo86, thanks for the details. Actually, in ginkgo v1, the BeforeSuite and AfterSuite can run successfully, so users can test the default case(CRDs, Controllers, and Webhooks can be installed successfully, just as what BeforeSuite does) without any changes just running make test. My point is to reserve the same adaption for users in the newest kubebuilder(v4). I don't know whether it makes sense to the community.
> 
> Can you please describe what problem and when are you facing it? If I create a new project (kubebuilder init) Scaffold an API (kubebuilder create api) Run (make test)
> 
> All works fine. So, can you pelase raise an issue with:
> 
> * What are the steps that you are performing
> * What would like to see
> * What are you checking instead?

Thanks for the advice, I just opened an issue that can reproduce the bug.
https://github.com/kubernetes-sigs/kubebuilder/issues/3646, FYI.

**Comment by dashanji** [comment]
Hi, @camilamacedo86 Could you please take another review? Thanks.

**Comment by Sajiyah-Salat** [comment]
@dashanji one thing that I noticed is common that there is a func() keywork everytwhere in the code. I tried to find its origin and the logic it contains. I didnt find it till now through ide. I know this is a sillly question. But can you tell me whats the logic wriiten in func() keyword. or guide me to its origin and I will find it by my own.

File: docs/book/src/cronjob-tutorial/testdata/project/api/v1/cronjob_webhook_test.go
Comment (camilamacedo86) [inline docs/book/src/cronjob-tutorial/testdata/project/api/v1/cronjob_webhook_test.go:27]: Why the code is commented?
We should scaffold only tests that can work

File: pkg/plugins/golang/v3/scaffolds/api.go
Comment (camilamacedo86) [inline pkg/plugins/golang/v3/scaffolds/api.go:102]: Can you please remove it from here?
the plugin go/v3 is deprecated so we should not perform changes under this one
Any change must be applied only in go/v4

File: testdata/project-v4/api/v1/firstmate_webhook_test.go
Comment (camilamacedo86) [inline testdata/project-v4/api/v1/firstmate_webhook_test.go:56]: Remove the comments and ensure that we have a generic scaffold that can be valid for all case scenarios.
In this case just replace those by `TODO(user): Add your logic here`

**Review by camilamacedo86 — COMMENTED** [review]
Great work 🥇 
Just a few nits:
- We should not scaffold a code that is not valid. So, for the webhooks under the test add a TODO(user): Add your logic here as we do for the reconcile and it is fine
- We should not change the go/v3 since it is deprecated. Can you please revert the changes and re-run make generate? 
- The title of the PR we use to generate the changelog, I will update it accordingly to try to help out but you can check more about it in the CONTRIBUTION.md guide

After the changes please ensure that we have only 1 commit as of now so that we can get this one merged. You can squash the commits.

**Comment by dashanji** [comment]
> @dashanji one thing that I noticed is common that there is a `func()` keywork everywhere in the code. I tried to find its origin and the logic it contains. I didnt find it till now through ide. I know this is a sillly question. But can you tell me whats the logic wriiten in `func()` keyword. or guide me to its origin and I will find it by my own.

Hi, @Sajiyah-Salat. A possible way is to debugging the `kubebuilder create api/webhook` then you can see the complete execution order.

The webhook test is generated [here](https://github.com/kubernetes-sigs/kubebuilder/pull/3631/files#diff-2b45b3690775d205ba84e67e27f8fbda705efb65adc71c8c6c87ade18c7ffe80R107) and its logic code is [here](https://github.com/kubernetes-sigs/kubebuilder/pull/3631/files#diff-978c31cb03415b62b55525d4ac25f74e9715997ccdd174ef6a0d58335d131620).

The controller test is generated [here](https://github.com/kubernetes-sigs/kubebuilder/pull/3631/files#diff-d811acb12d570acb9a4162ac14752d9a41edd82926e01e65eaf5ba4e533ee010R90) and its logic code is [here](https://github.com/kubernetes-sigs/kubebuilder/pull/3631/files#diff-978c31cb03415b62b55525d4ac25f74e9715997ccdd174ef6a0d58335d131620). FYI.

File: pkg/plugins/golang/v3/scaffolds/api.go
Comment (dashanji) [reply pkg/plugins/golang/v3/scaffolds/api.go:102]: Ok.

**Comment by dashanji** [comment]
> Great work 🥇 Just a few nits:
> 
> * We should not scaffold a code that is not valid. So, for the webhooks under the test add a TODO(user): Add your logic here as we do for the reconcile and it is fine
> * We should not change the go/v3 since it is deprecated. Can you please revert the changes and re-run make generate?
> * The title of the PR we use to generate the changelog, I will update it accordingly to try to help out but you can check more about it in the CONTRIBUTION.md guide
> 
> After the changes please ensure that we have only 1 commit as of now so that we can get this one merged. You can squash the commits.

Make sense to me.

File: docs/book/src/cronjob-tutorial/testdata/project/api/v1/cronjob_webhook_test.go
Comment (dashanji) [reply docs/book/src/cronjob-tutorial/testdata/project/api/v1/cronjob_webhook_test.go:27]: These codes are just a reference sample for users, which makes it easier for them to get started quickly. However, they can work actually. Do you think we should delete all of this?

File: testdata/project-v4/api/v1/firstmate_webhook_test.go
Comment (dashanji) [reply testdata/project-v4/api/v1/firstmate_webhook_test.go:56]: ok

File: docs/book/src/cronjob-tutorial/testdata/project/api/v1/cronjob_webhook_test.go
Comment (camilamacedo86) [reply docs/book/src/cronjob-tutorial/testdata/project/api/v1/cronjob_webhook_test.go:27]: I think we need to keep the project standard and only scaffold code that works
in this case, if we add a TODO(user): Add here your logic it is fine because it is == we do in the other places.

Comment (dashanji) [reply docs/book/src/cronjob-tutorial/testdata/project/api/v1/cronjob_webhook_test.go]: Done.

**Review by camilamacedo86 — APPROVED** [review]
It shows great for me
I am approving 

But would be nice see if we get a second reviewer as well
Let's try, if nobody be able to do this review I can either lgtm after a while

**Review by camilamacedo86 — APPROVED** [review]
@varshaprasad96 @everettraven @Kavinjsir 

It seems OK for me. 
Since it is open for a while I will /lgtm 
otherwise the author might need to rebase this one

However, if you have suggestions/objections and etc we can always address in a follow up.

**Review by Kavinjsir — APPROVED** [review]
Looks good to me!

**Pull request merged**
